osu.Game.Tournament/IPC/FileBasedIPC.cs

```diff
@@ -184,29 +180,28 @@ public bool SetIPCLocation(string path)
         /// <returns>Whether an IPC directory was successfully auto-detected.</returns>
         public bool AutoDetectIPCLocation() => SetIPCLocation(findStablePath());
 
-        private static bool ipcFileExistsInDirectory(string p) => p != null && File.Exists(Path.Combine(p, "ipc.txt"));
+        private static bool ipcFileExistsInDirectory(string? p) => p != null && File.Exists(Path.Combine(p, "ipc.txt"));
 
-        [CanBeNull]
-        private string findStablePath()
+        private string? findStablePath()
         {
-            string stableInstallPath = findFromEnvVar() ??
-                                       findFromRegistry() ??
-                                       findFromLocalAppData() ??
-                                       findFromDotFolder();
+            string? stableInstallPath = findFromEnvVar() ??
+                                        findFromRegistry() ??
+                                        findFromLocalAppData() ??
+                                        findFromDotFolder();
 
             Logger.Log($"Stable path for tourney usage: {stableInstallPath}");
             return stableInstallPath;
         }
 
-        private string findFromEnvVar()
+        private string? findFromEnvVar()
         {
             try
             {
                 Logger.Log("Trying to find stable with environment variables");
-                string stableInstallPath = Environment.GetEnvironmentVariable("OSU_STABLE_PATH");
+                string? stableInstallPath = Environment.GetEnvironmentVariable("OSU_STABLE_PATH");
 
                 if (ipcFileExistsInDirectory(stableInstallPath))
-                    return stableInstallPath;
+                    return stableInstallPath!;
```

Comment:
isn't necessary if you annotate `ipcFileExistsInDirectory()` accordingly

```diff
diff --git a/osu.Game.Tournament/IPC/FileBasedIPC.cs b/osu.Game.Tournament/IPC/FileBasedIPC.cs
index 333dd0fd73..daf1417f6f 100644
--- a/osu.Game.Tournament/IPC/FileBasedIPC.cs
+++ b/osu.Game.Tournament/IPC/FileBasedIPC.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 using System.IO;
 using System.Linq;
 using Microsoft.Win32;
@@ -180,7 +181,7 @@ public bool SetIPCLocation(string? path)
         /// <returns>Whether an IPC directory was successfully auto-detected.</returns>
         public bool AutoDetectIPCLocation() => SetIPCLocation(findStablePath());
 
-        private static bool ipcFileExistsInDirectory(string? p) => p != null && File.Exists(Path.Combine(p, "ipc.txt"));
+        private static bool ipcFileExistsInDirectory([NotNullWhen(true)] string? p) => p != null && File.Exists(Path.Combine(p, "ipc.txt"));
 
         private string? findStablePath()
         {
@@ -201,7 +202,7 @@ public bool SetIPCLocation(string? path)
                 string? stableInstallPath = Environment.GetEnvironmentVariable("OSU_STABLE_PATH");
 
                 if (ipcFileExistsInDirectory(stableInstallPath))
-                    return stableInstallPath!;
+                    return stableInstallPath;
             }
             catch
             {

```

Reply:
ok